### ISSUE TITLE
os.ExitError cannot be mocked

### ISSUE THREAD

**Issue by dbuteyn** [golang/go#58988]
Our application happens to make a lot of shell calls. Up until now we've essentially mocked out a large portion of the application, allowing `go test` to proceed as expected.

Simulating a shell command isn't particularly hard, all one needs to do is write a function that generates the same output. Problems appear when process exit codes are being tested.

Under normal circumstances `Cmd.Run()` will return an `*os.ExitError`. Getting the process exit code requires one to either do `err.(*os.ExitError).Sys().(syscall.WaitStatus).ExitStatus()` (<1.12) or `err.(*os.ExitError).ExitCode()` (<1.13) or `errors.As(err, <*os.ExitError>)`. This means that the simulated error _must_ be a `*os.ExitError` or these type assertion will fail.

Yes, I can (and must) refactor our code so that a simulated exit code can be returned. Would have been nice however if os.ExitError was an interface. I'm not even sure why by design its so hard to get process exit codes, especially considering Linux, Windows and Mac all have them.

**Comment by seankhliao** [comment]
we can't change it to be an interface.

**Comment by dbuteyn** [comment]
Let me rephrase as a question then, what is the idiomatic way of mocking `Cmd.Run()` and friends? I ended up writing a wrapper and returning a custom error type and rewriting every single line that got anywhere near "os/exec".

I'm not asking to change os.ExitError to an interface, that kind of change will likely break every single Go application in existence. I did however see Go 2.0 floating around somewhere. Breaking changes are usually allowed/expected in a major version update.

**Comment by ianlancetaylor** [comment]
There are no plans for Go 2.0, and it is very unlikely.  More possible would be an os/exec/v2 package, but there are no plans for that either.

It should be feasible to write an interface type with `Sys` and `ExitCode` methods, and use that in the code that you want to mock.  But that is a discussion for golang-nuts or some other forum, not here.  See https://go.dev/wiki/Questions.

**Comment by samherrmann** [comment]
If you are just interested in the exit code, then you can create your own interface to stub out `*exec.ExitError`.

Interface:
```go
type ExitError interface {
	error
	ExitCode() int
}
```

Stub:
```go
// ExitErrorStub is a stub for exec.ExitError.
type ExitErrorStub struct {
	code int
}

// Error implements the error interface.
func (e *ExitErrorStub) Error() string {
	return fmt.Sprintf("exit code %v", e.code)
}

// ExitCode implements the ExitError interface.
func (e *ExitErrorStub) ExitCode() int {
	return e.code
}

// Ensure that *exitErrorStub implements the exitError interface.
var _ ExitError = (*ExitErrorStub)(nil)
```

Example function making use of `ExitError`:
```go
// extractExitCode returns the exit code from err if err is an exit error. -1 is
// returned if err is not an exit error.
func extractExitCode(err error) int {
	var exitErr ExitError
	if errors.As(err, &exitErr) {
		return exitErr.ExitCode()
	}
	return -1
}
```

Test function:
```go
func Test_extractExitCode(t *testing.T) {
	t.Run("using stub", func(t *testing.T) {
		err := &ExitErrorStub{code: 3}
		got := extractExitCode(err)
		want := 3
		if got != want {
			t.Fatalf("got %v, want %v", got, want)
		}
	})

	t.Run("using real exit error", func(t *testing.T) {
		cmd := exec.Command("sh", "-c", "exit 3")
		_, err := cmd.CombinedOutput()
		if err == nil {
			t.Fatal("expected error but didn't get one")
		}

		got := extractExitCode(err)
		want := 3
		if got != want {
			t.Fatalf("got %v, want %v", got, want)
		}
	})
}
```

Run this on [Go Playground](https://go.dev/play/p/H0LK4Kqz8u6)

I hope this will help someone.

**Comment by dbuteyn** [comment]
@samherrmann yes, that's pretty much [what I described](https://github.com/golang/go/issues/58988#issuecomment-1465197764) doing a couple months ago. But as you've noticed yourself it is a lot of boilerplate for something that should have been trivial.

Preferably cmd.Run() and friends would simply return the exit code as a third return value, thus also closing this issue since there's no need to mock os.ExitError. For reasons however the exit code got lodged deep inside the error itself.

But no such interface-breaking changes are planned so... one can only dream.
